### PR TITLE
[SPARK-28762][CORE] Read JAR main class if JAR is not located in local file system

### PR DESCRIPTION
**Pull request by igozali**
### What changes were proposed in this pull request?

JIRA: https://issues.apache.org/jira/browse/SPARK-28762

TL;DR: Automatically read the `Main-Class` from a JAR's manifest even if the JAR isn't in the local file system (i.e. in S3 or HDFS).

### Why are the changes needed?

When deploying a fat JAR (e.g. using `sbt-assembly`) to S3/HDFS, users might choose to include the main class for the JAR in its manifest. This change allows the user to `spark-submit` the JAR without having to specify the main class again via the `--class` argument.

### Does this PR introduce any user-facing change?

Yes. Previously, if the primary resource is a JAR and isn't located in the local file system, it will fail with the error:
```
$ spark-submit s3a://nonexistent.jar
Exception in thread "main" org.apache.spark.SparkException: Cannot load main class from JAR s3a://nonexistent.jar with URI s3a. Please specify a class through --class.
    ...
```

With this PR, the main class will be read from the manifest, assuming the classpath contains the appropriate JAR to read the file system.

### How was this patch tested?

Added some tests in `core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala`.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
Thank you for making a PR, @igozali

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by igozali** [comment]
No problem, thanks for taking a look @dongjoon-hyun. I was wondering if the approach is ok and whether there are any other test cases I should add? Right now, I have a few for S3 using the fake S3 file system test utility.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 1]
```diff
@@ -213,24 +215,19 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
...
-          } catch {
-            case _: Exception =>
-              error(s"Cannot load main class from JAR $primaryResource")
-          }
-        case _ =>
-          error(
-            s"Cannot load main class from JAR $primaryResource with URI $uriScheme. " +
-            "Please specify a class through --class.")
+      try {
+        val fs = FileSystem.get(
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: To avoid having to involve Spark/Hadoop config etc in the existing code path for "file", how about just doing this part in the 'else' case of the current match? It may be just fine this way too, just might be more conservative. I don't feel too strongly about it, as long as nothing breaks current functionality.

**Review by srowen — CHANGES_REQUESTED** [review]

**Comment by igozali** [comment]
Hi @srowen, thanks for taking the time to give a review. For your last comment, if I try to keep as much of the original structure, there will be a bit of superfluous code for the error being thrown, so I moved the try-catch outside the uriScheme match - I hope that's ok.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 2]
```diff
@@ -216,20 +218,27 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
...
-            case _: Exception =>
-              error(s"Cannot load main class from JAR $primaryResource")
-          }
-        case _ =>
+          case _ =>
+            val sparkConf = new SparkConf()
+            sparkConf.setAll(sparkProperties)
+            val fs = FileSystem.get(uri, SparkHadoopUtil.newConfiguration(sparkConf))
+
+            Utils.tryWithResource(new JarInputStream(fs.open(new Path(primaryResource)))) { jar =>
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: When submitting in client mode, this will cause a duplicate read of the remote jar in the driver process, because the `SparkSubmit` class also does the download to make it available to the driver.

I think it would be better to move this whole "figure out the main class if it's not defined" logic to `SparkSubmit` after that is done (around L370, look for `var localPrimaryResource: String`).

In cluster mode you'll still need code similar to this to read the manifest, but that should be easy to handle.

Comment (igozali) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: Thanks, that's really valuable input. I'm not a big fan of my own approach of having Hadoop classes being used in `SparkSubmitArguments` anyway. I'll take a look into this, but probably won't be able to get into it until this weekend.

Comment (srowen) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: @igozali have you had a chance to take another look at this?

Comment (igozali) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: Hi @srowen My bad, I dropped the ball here - got caught up in work. Still interested in this though, and will take a look at it before tomorrow night.

**Comment by igozali** [comment]
@srowen @vanzin I've updated my approach, would like to see if this is more along the lines of what you expect.

File: core/src/main/scala/org/apache/spark/TestUtils.scala [hunk 3]
```diff
@@ -97,9 +97,20 @@ private[spark] object TestUtils {
...
    */
-  def createJar(files: Seq[File], jarFile: File, directoryPrefix: Option[String] = None): URL = {
+  def createJar(
+      files: Seq[File],
+      jarFile: File,
+      directoryPrefix: Option[String] = None,
+      mainClass: Option[String] = None): URL = {
+    val manifest = mainClass match {
+      case Some(mc) => new Manifest(
+        new ByteArrayInputStream(s"Manifest-Version: 1.0\nMain-Class: $mc\n".getBytes)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/TestUtils.scala]: Use `m.getMainAttributes().putValue(...)` instead?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 4]
```diff
@@ -203,6 +204,52 @@ private[spark] class SparkSubmit extends Logging {
...
+    args: SparkSubmitArguments,
+    hadoopConf: HadoopConfiguration,
+    jarPath: String
+  ): String = {
+    if (args.mainClass != null) {
+      return args.mainClass
+    }
+
+    var mainClass: String = null
+    if (args.mainClass == null && !args.isPython && !args.isR && jarPath != null) {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Checking `args.mainClass` is redundant here. It also feels like the other checks for R and Python don't really belong here.

This method would be better if all it did was get the main class from a jar file (regardless of whether the file is local or remote).

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 5]
```diff
@@ -437,6 +484,22 @@ private[spark] class SparkSubmit extends Logging {
       }.orNull
     }
 
+    // At this point, we have attempted to download all remote resources.
+    // Now we try to resolve the main class if our primary resource is a JAR.
+    // First, try the localPrimaryResource.
+    if (args.mainClass == null) {
+      args.mainClass = resolveMainClassIfNeeded(args, hadoopConf, localPrimaryResource)
+    }
+    // If we can't get it from localPrimaryResource, use whatever the original argument was.
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: In what case is `localPrimaryResource` not a local copy of `args.primaryResource`?

If seems to me that what you want is:

```
if (mainClass == null && not python && not R) {
  mainClass = readMainClass(Option(localPrimaryResource).getOrElse(args.primaryResource), hadoopConf)
  if (mainClass == null) yellAtUser()
}
```

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala [hunk 6]
```diff
@@ -453,6 +453,86 @@ class SparkSubmitSuite
...
+        appArgs, conf = Some(hadoopConf))
+
+      mainClass_ should be ("SomeMainClass")
+      classpaths should have length 1
+      classpaths.head should endWith (jarFile.getName)
+      childArgs.mkString(" ") should be ("arg1 arg2")
+    }
+  }
+
+  test("automatically sets mainClass if primary resource is S3 JAR in cluster mode") {
```
Comment (vanzin) [inline core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala]: This is basically the same as above. Do:

```
Seq("client", "cluster").foreach { mode =>
  test(s"blah in $mode mode") {
    callTestMethodHere()
  }
}
```

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 7]
```diff
@@ -203,6 +204,52 @@ private[spark] class SparkSubmit extends Logging {
...
+
+      try {
+        uriScheme match {
+          case "file" =>
+            // If local file, probably more stable to use JarFile than Hadoop's FileSystem class
+            Utils.tryWithResource(new JarFile(uri.getPath)) { jar =>
+              mainClass = jar.getManifest.getMainAttributes.getValue("Main-Class")
+            }
+          case _ =>
+            val fs = FileSystem.get(uri, hadoopConf)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Also, you could use this code for both local and remote files, right?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 5 above)
Comment (igozali) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: > In what case is `localPrimaryResource` not a local copy of `args.primaryResource`?

I'm not quite sure actually, but right above this code is a `clusterManager == YARN` check where it seems to be using `args.primaryResource` instead of `localPrimaryResource`, so I just tried to check both variables?

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala (hunk 6 above)
Comment (igozali) [reply core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala]: The assertions for both tests are slightly different, though. I will try deduplicating the common logic in both tests.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 8]
```diff
@@ -437,6 +438,37 @@ private[spark] class SparkSubmit extends Logging {
       }.orNull
     }
 
+    def resolveMainClassIfNeeded(jarPath: String): Try[String] = {
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: I think this looks better inlined below in the only place where it's called; also using a normal `try...catch` instead of a `Try` + match.

**Comment by vanzin** [comment]
Merging to master.

**Comment by igozali** [comment]
Thanks @vanzin @srowen @dongjoon-hyun for all of your help!
